Unable to scale in due to mismatch between hostname and instance name

Hi! Thank you for all the work done to build this plugin.

I'm trying to setup the autoscaler on Infomaniak Public Cloud to run batch workloads.

I got it working up to the point when jobs complete and we need to scale in the pool and then it simply breaks. Here is sample of the log from the autoscaler:

```
2024-07-13T11:02:41.302Z [INFO]  policy_eval.worker: scaling target: id=96c185b6-63cc-2a8d-94a8-7220e0a5027f policy_id=a71493ba-9f09-1d6b-4eea-3f7ffd226b9e queue=cluster target=os-nova from=19 to=2 reason="scaling down because factor is 0.079576" meta=map[nomad_policy_id:a71493ba-9f09-1d6b-4eea-3f7ffd226b9e]
2024-07-13T11:02:41.628Z [DEBUG] external_plugin.os-nova: performing node pool filtering: combined_identifier="node_class:batch and datacenter:infomaniak-dc3 and node_pool:batch" timestamp=2024-07-13T11:02:41.628Z
2024-07-13T11:02:41.638Z [DEBUG] external_plugin.os-nova: found node: draining=false eligibility=eligible node_pool=batch status=ready datacenter=infomaniak-dc3 node_class=batch node_id=063c0c50-b038-a041-1929-24de46c317ad timestamp=2024-07-13T11:02:41.637Z
2024-07-13T11:02:41.638Z [DEBUG] external_plugin.os-nova: found node: status=ready datacenter=infomaniak-dc3 node_class=batch node_pool=batch draining=false eligibility=eligible node_id=31bd84f4-a29c-a105-cf24-395751fd4a66 timestamp=2024-07-13T11:02:41.637Z
[...]
2024-07-13T11:02:41.639Z [DEBUG] external_plugin.os-nova: node passed filter criteria: node_id=063c0c50-b038-a041-1929-24de46c317ad timestamp=2024-07-13T11:02:41.638Z
2024-07-13T11:02:41.639Z [DEBUG] external_plugin.os-nova: node passed filter criteria: node_id=31bd84f4-a29c-a105-cf24-395751fd4a66 timestamp=2024-07-13T11:02:41.638Z
[...]
2024-07-13T11:02:41.644Z [DEBUG] external_plugin.os-nova: identified remote provider ID for node: node_id=063c0c50-b038-a041-1929-24de46c317ad remote_id=nomad-batch-ff9c8fc9-7dd4.dc3-a.pub1.infomaniak.cloud timestamp=2024-07-13T11:02:41.643Z
2024-07-13T11:02:41.649Z [DEBUG] external_plugin.os-nova: identified remote provider ID for node: node_id=31bd84f4-a29c-a105-cf24-395751fd4a66 remote_id=nomad-batch-2e25d4d4-b5ab.dc3-a.pub1.infomaniak.cloud timestamp=2024-07-13T11:02:41.649Z
[...]
2024-07-13T11:02:41.752Z [ERROR] policy_eval.worker: failed to evaluate policy: eval_id=e8ef91d0-3cf6-d2bb-b706-51d958c65504 eval_token=7ce95066-412d-087c-374c-7a3b77c930fc id=96c185b6-63cc-2a8d-94a8-7220e0a5027f policy_id=a71493ba-9f09-1d6b-4eea-3f7ffd226b9e queue=cluster error="failed to scale target: rpc error: code = Unknown desc = failed to perform scaling action: failed to perform pre-scale Nomad scale in tasks: no nodes identified for scaling in action"
```

What seems to happen here is that on Infomaniak cloud, the `unique.platform.aws.hostname` attribute is a full hostname (like `nomad-batch-e0a3f0ee-5418.dc3-a.pub1.infomaniak.cloud`) and not simply the instance name from Openstack.

- One one hand, [`osNovaNodeIDMapBuilder`](https://github.com/jorgemarey/nomad-nova-autoscaler/blob/9f5b0dea51095004517bc5b295fbd73a45ca355f/plugin/openstack.go#L745) is picking `unique.platform.aws.hostname` as the "ClusterNodeID" for [`IdentifyScaleInRemoteIDs`](https://github.com/hashicorp/nomad-autoscaler/blob/b12b2884df7067f72b7203286b411733d48e608c/sdk/helper/scaleutils/cluster.go#L270) [[setup here](https://github.com/jorgemarey/nomad-nova-autoscaler/blob/9f5b0dea51095004517bc5b295fbd73a45ca355f/plugin/plugin.go#L123)], ending up with `nomad-batch-e0a3f0ee-5418.dc3-a.pub1.infomaniak.cloud`
- Meanwhile, [`countServers`](https://github.com/jorgemarey/nomad-nova-autoscaler/blob/9f5b0dea51095004517bc5b295fbd73a45ca355f/plugin/openstack.go#L469) is picking the instance name from OS, ending up with `nomad-batch-e0a3f0ee-5418`

Then, [`RunPreScaleInTasksWithRemoteCheck`](https://github.com/hashicorp/nomad-autoscaler/blob/b12b2884df7067f72b7203286b411733d48e608c/sdk/helper/scaleutils/cluster.go#L145) gets confused because `id.RemoteResourceId` is the full hostname while `remoteId` is the instance name. Everything is filtered out, and nothing is left to scale in.

I only have the following attributes to work with:
| Attribute | Value |
| --- | --- |
| unique.hostname | `nomad-batch-e0a3f0ee-5418` |
| unique.network.ip-address | `10.0.0.144` |
| unique.platform.aws.hostname | `nomad-batch-e0a3f0ee-5418.dc3-a.pub1.infomaniak.cloud` |
| unique.platform.aws.instance-id | `i-001787c3` (doesn't seem to match anything from `openstack server show ...`)  |
| unique.platform.aws.public-hostname | `nomad-batch-e0a3f0ee-5418.dc3-a.pub1.infomaniak.cloud` |

Notably, nothing has the instance ID.

I tried to set `id_attribute = "unique.hostname"` as this one is indeed the instance name. But setting anything as `id_attribute` also sets [`t.idMapper = true`](https://github.com/jorgemarey/nomad-nova-autoscaler/blob/9f5b0dea51095004517bc5b295fbd73a45ca355f/plugin/plugin.go#L124), forcing me to select the instance ID and not its name. 🙃

For completion's sake, here is the configuration for the plugin:
```hcl
# autoscaler config
target "os-nova" {
  driver = "os-nova"
  config = {
    auth_url     = "https://api.pub1.infomaniak.cloud/identity"
    username     = "..."
    password     = "..."
    domain_name  = "Default"
    project_id   = "..."
    project_name = "..."
    region_name  = "dc3-a"
    id_attribute = "unique.hostname"
  }
}

# policy
target "os-nova" {
  dry-run = false

  evenly_split_azs    = true
  stop_first          = true
  image_name          = "Fedora Core OS 40"
  flavor_name         = "a4-ram16-disk20-perf1"
  pool_name           = "nomad-autoscaler-batch"
  name_prefix         = "nomad-batch-"
  network_id          = "..."
  user_data_template  = "local/user_data"

  datacenter                    = "infomaniak-dc3"
  node_class                    = "batch"
  node_pool                     = "batch"
  node_drain_deadline           = "1h"
  node_drain_ignore_system_jobs = false
  node_purge                    = true
  node_selector_strategy        = "empty_ignore_system"
}
```

I don't think there is any solution to the issue with only configuration changes. I am open to submit a pull request, but what would be the prefered way to tackle this?

Thanks

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,9 +1,15 @@
+## 0.4.2 (Jul 15, 2024)
+
+FEATURES:
+* Add `name_attribute` to specify the attribute to use when selecting the instance name to 
+search for the instance to downscale
+
 ## 0.4.1 (Jun 27, 2024)
 
 FEATURES:
 * Updated golang to 1.22.4
 * Updated autoscaler dependendy to 0.4.4
-* floatingip_pool_name can now be specified to attach a floating ip for created instances
+* `floatingip_pool_name` can now be specified to attach a floating ip for created instances
 * fix cache data being renoved after reload
 
 ## 0.4.0 (Nov 23, 2023)

--- a/README.md
+++ b/README.md
@@ -37,8 +37,8 @@ target "os-nova" {
 * `cacert_file` `(string: "")` - Location of the certificate to use for OS APIs verification
 * `insecure_skip_verify` `(string: "")` - Skip TLS certificate verification
 
-* `id_attribute` `(string: "")` - The nomad attribute to use that maps the nomad client to an OS Compute instance. By default `unique.platform.aws.hostname` is used and a previous search is needed
-to get the instance id using the instance name
+* `name_attribute` `(string: "unique.platform.aws.hostname")` - The nomad attribute that reflects the instance name. This needs to be used for searching the instance ID in the proccess of downscaling
+* `id_attribute` `(string: "")` - The nomad attribute to use that maps the nomad client to an OS Compute instance. If not specified then a previous search is needed to get the instance id using the instance name using `name_attribute`. If this is specified it takes priority over `name_attribute`
 * `action_timeout` `(string: "")` - The timeout to use when performing create and delete actions over servers. This should be specified as a duration. The default vaule is 90s
 * `ignored_states` `(string: "")` - A comma-separated list of server states to be ignored. The complete list can be seen [here](https://docs.openstack.org/api-guide/compute/server_concepts.html)
 

--- a/plugin/openstack.go
+++ b/plugin/openstack.go
@@ -30,13 +30,14 @@ import (
 )
 
 const (
-	version = "v0.4.1"
+	version = "v0.4.2"
 )
 
 const (
 	defaultActionTimeout        = 120 * time.Second
 	defaultStatusTImeout        = 5 * time.Minute
 	defaultScaleTimeout         = 2 * time.Hour
+	defaultNameProperty         = "unique.platform.aws.hostname"
 	poolTag                     = "na_pool:%s"
 	defaultConfigValueSeparator = ","
 	configKVSeparator           = "="
@@ -742,10 +743,14 @@ func (t *TargetPlugin) getInstancePortID(id string) (string, error) {
 
 // osNovaNodeIDMapBuilder is used to identify the Opensack Nova ID of a Nomad node using
 // the relevant attribute value.
-func osNovaNodeIDMapBuilder(property string) scaleutils.ClusterNodeIDLookupFunc {
+func osNovaNodeIDMapBuilder(nameProperty, idProperty string) scaleutils.ClusterNodeIDLookupFunc {
 	var isMeta bool
-	if property == "" {
-		property = "unique.platform.aws.hostname"
+	property := defaultNameProperty
+	if nameProperty != "" {
+		property = nameProperty
+	}
+	if idProperty != "" {
+		property = idProperty
 	}
 	if strings.HasPrefix(property, "meta.") {
 		isMeta = true

--- a/plugin/plugin.go
+++ b/plugin/plugin.go
@@ -29,7 +29,8 @@ const (
 	configKeyCACertFile  = "cacert_file"
 	configKeyInsecure    = "insecure_skip_verify"
 
-	configKeyNodeIDAttr = "id_attribute"
+	configKeyNodeIDAttr   = "id_attribute"
+	configKeyNodeNameAttr = "name_attribute"
 
 	configKeyName           = "name"
 	configKeyNamePrefix     = "name_prefix"
@@ -120,7 +121,7 @@ func (t *TargetPlugin) SetConfig(config map[string]string) error {
 
 	// Store and set the remote ID callback function.
 	t.clusterUtils = clusterUtils
-	t.clusterUtils.ClusterNodeIDLookupFunc = osNovaNodeIDMapBuilder(config[configKeyNodeIDAttr])
+	t.clusterUtils.ClusterNodeIDLookupFunc = osNovaNodeIDMapBuilder(config[configKeyNodeNameAttr], config[configKeyNodeIDAttr])
 	t.idMapper = config[configKeyNodeIDAttr] != ""
 
 	return nil

